### ISSUE TITLE
Add LSP client flag to pull diagnostics on save

### ISSUE THREAD

**Issue by Delerme** [neovim/neovim#29792]
### Problem

Now that neovim 10 supports pull diagnostics, I think it'd be useful to expose some additional flags to control how often diagnostic requests are sent.

Right now, diagnostics are requested after each `didChange` event, which can be controlled partially via existing lsp client flags like `debounce_text_changes`: https://neovim.io/doc/user/lsp.html#_lua-module:-vim.lsp.client

Some lsp servers like eslint expose options to run diagnostics on save, since they can be a little slow. [With pull diagnostics, those settings have to be on the client-side to function](https://github.com/neovim/nvim-lspconfig/issues/3211).

### Expected behavior

To start, I propose adding a new flag `pull_diagnostics_on_save` that sends diagnostic requests after `didSave` instead of `didChange` for that lsp client.

Any thoughts or concerns? I'm happy to try implementing if it sounds like a reasonable idea!

**Comment by Delerme** [comment]
cc @catlee for any thoughts or guidance, or to catch any mistakes in my understanding. Thanks for adding pull diagnostic support!

**Comment by catlee** [comment]
Yes, I think it makes sense to be able to configure the frequency of when diagnostics are requested per client or to change it to be on save only.

According to https://github.com/neovim/neovim/blob/7031949be065870f0daf74a9f1be3df47f83312c/runtime/doc/lsp.txt#L1027, you can already configure the debounce timer per client.

**Comment by MariaSolOs** [comment]
This sounds like a good idea. @Delerme feel free to start working on a draft implementation!

**Comment by MariaSolOs** [comment]
@justinmk @mfussenegger based on the conversation from https://github.com/neovim/neovim/pull/30055, it seems like we've agreed that this behavior can be achieved today by calling `vim.diagnostic.enable` in autocommands. I guess we can close this?

**Comment by thatsmydoing** [comment]
I don't believe that's the case. I think what they were referring to is `vim.lsp.diagnostic._enable` which is accessible but it's technically private. `vim.diagnostic.enable` just turns off the display of the diagnostics but doesn't prevent them from being requested (which is what we want to stop).

**Comment by liskin** [comment]
Has there been any progress on this one? I recently started using ruff via https://github.com/neovim/nvim-lspconfig/blob/master/lsp/ruff.lua and the diagnostics are often delayed several seconds because of my setting `debounce_text_changes = 5000`. (Note that one has to either disable `semantic_tokens` and `document_color` or apply [a workaround](https://github.com/neovim/neovim/issues/39785#issuecomment-4450886150) for `debounce_text_changes` to actually debounce)

pylsp isn't affected by this because nvim sends `didSave` to it, but no `didSave` is being sent to ruff (only the debounced `didChange`) and diagnostics are only being pulled after the debounced `didChange`.

Is there any way I can force a `didChange`/`didSave` to be sent when the buffer is saved?

**Comment by liskin** [comment]
Actually, I figured it out myself (it was extra annoying indeed so I had to lol):

```lua
-- force sending didChange on save to servers that don't ask for didSave
vim.api.nvim_create_autocmd('BufWritePost', {
	group = vim.api.nvim_create_augroup('DidChangeFlush', {}),
	callback = function(ev)
		for _, client in pairs(vim.lsp.get_clients({ bufnr = ev.buf })) do
			require("vim.lsp._changetracking").flush(client, ev.buf)
		end
	end,
})
```

It's clearly a hack as `vim.lsp._changetracking` isn't public API but it seems to work with both 0.11 and 0.12 so that's good for now.

**Comment by liskin** [comment]
Some extra context…

didSave autocmd: https://github.com/neovim/neovim/blob/6780e78b0981f52aec9c4ef96356d3fef3742c9a/runtime/lua/vim/lsp.lua#L939-L946

didSave impl that only sends didSave if the server asked for it: https://github.com/neovim/neovim/blob/6780e78b0981f52aec9c4ef96356d3fef3742c9a/runtime/lua/vim/lsp/_changetracking.lua#L207-L219

how the didChange happens before the didSave: https://github.com/neovim/neovim/blob/6780e78b0981f52aec9c4ef96356d3fef3742c9a/runtime/lua/vim/lsp/client.lua#L851-L854

My autocmd simply unconditionally calls `flush` always to make sure `didChange` gets sent on save regardless of whether `didSave` got sent or not. Seems to me this might be a good thing to do anyway? Shall I open a PR?

**Comment by liskin** [comment]
> My autocmd simply unconditionally calls `flush` always to make sure `didChange` gets sent on save regardless of whether `didSave` got sent or not. Seems to me this might be a good thing to do anyway? Shall I open a PR?

Something like this might do:

```diff
diff --git ./runtime/lua/vim/lsp/_changetracking.lua ./runtime/lua/vim/lsp/_changetracking.lua
index 0788cd186d..a8cbcf856c 100644
--- ./runtime/lua/vim/lsp/_changetracking.lua
+++ ./runtime/lua/vim/lsp/_changetracking.lua
@@ -216,6 +216,8 @@ function M._send_did_save(bufnr)
           },
           text = included_text,
         })
+      else
+        M.flush(client, bufnr)
       end
     end
   end
```

**Comment by justinmk** [comment]
> Something like this might do:

@Emilv2 @igorlfs WDYT ?

**Comment by igorlfs** [comment]
Sorry, can't offer much insight here.

**Comment by justinmk** [comment]
@liskin your patch is simple enough, if you want to send a PR we can just try it and revert if something goes wrong.

**Comment by MariaSolOs** [comment]
> [@liskin](https://github.com/liskin) your patch is simple enough, if you want to send a PR we can just try it and revert if something goes wrong.

Implementing this [here](https://github.com/neovim/neovim/pull/39894) (cc @justinmk)

**Comment by liskin** [comment]
> > [@liskin](https://github.com/liskin) your patch is simple enough, if you want to send a PR we can just try it and revert if something goes wrong.
>
> Implementing this here (cc [@justinmk](https://github.com/justinmk))

Oh, sorry for not following up. I haven't had a chance to open my computer since Friday as I'm on holiday. Patch is simple but I thought I was expected to add a unit test so I was going to do that during my travel tomorrow.

**Comment by MariaSolOs** [comment]
> > > [@liskin](https://github.com/liskin) your patch is simple enough, if you want to send a PR we can just try it and revert if something goes wrong.
> > 
> > 
> > Implementing this here (cc [@justinmk](https://github.com/justinmk))
> 
> Oh, sorry for not following up. I haven't had a chance to open my computer since Friday as I'm on holiday. Patch is simple but I thought I was expected to add a unit test so I was going to do that during my travel tomorrow.

No worries! Feel free to add the test to https://github.com/neovim/neovim/pull/39894 or open a new PR if you prefer :)
